### ISSUE TITLE
UNDEFINED-FUNCTION error when loading Sketch

### ISSUE THREAD

**Issue by pamoroso** [Interlisp/medley#1075]
**Describe the bug**
When loading the file `SKETCH` and answering yes to the question on whether `TEXTBOX` refers to `TEXTOBJ`, a break window with the error `UNDEFINED-FUNCTION` pops up.

**To Reproduce**
1. sign into your Interlisp Online account
2. under `Initial Exec`, select `Interlisp`
3. click `Run Medley`
4. at the Exec, evaluate `(LOAD 'SKETCH)`
5. when Exec fills the page and pauses, continue
6. when asked `TEXTBOX {in SK.SET.RECORD.LENGTHS} -> TEXTOBJ ?`, type `Y`

**Expected behavior**
Loading the file completes with no break window or errors.

**Screenshots**
The break window:
![SketchError1](https://user-images.githubusercontent.com/10342479/217918081-7f51dd5d-9ac9-4a3c-b242-0cd4e40051cb.png)

The full desktop:
![SketchError2](https://user-images.githubusercontent.com/10342479/217918074-f0622da7-9fe7-46d8-be99-9be031a46f18.png)

**Context (please complete the following information):**
 - OS: chromeOS Stable
 - OS Version: 109.0.5414.125
 - Host arch: x86_64 (ASUS Chromebox 3)
 - Maiko version: not sure how to get it
 - IL:MAKESYSDATE: 18-Dec-2022 20:59:08

**Additional context**
A break window with the same error pops up also when answering no to the question in step 6.

I haven't found much information on how to load and run Sketch, so I'm not sure loading the file `SKETCH` is one of the required steps and whether I'm going through the procedure in the proper order. So the observed behavior may or may not be an issue.

**Comment by rmkaplan** [comment]
try (LOAD ’SKETCH.LCOM) (the compiled file).  Could be that files with other compile-time declarations are needed if you load the source.

**Comment by pamoroso** [comment]
`(LOAD ’SKETCH.LCOM)` completes successfully and lets me run Sketch with `(SKETCH)`. It the tool isn't supposed to be loaded from source, this issue may be closed.

**Comment by rmkaplan** [comment]
Generally, sources should be loaded only if you want to work on them, not in them.  

Individual system source files (on sources/) typically have declarations that they want to make available for use by other system files.  They declare those by wrapping them in the EXPORT file package command.  Then, whenever a system is built (loadup-all.sh), there is a separate pass that scans all of the export-containing files and copies the declarations into another file, exports.all (also, the whereis database is created too).

So, to work on a file that really does depend on low-level stuff (and library/ files like SKETCH and TEDIT are allowed to do that), the first thing to do is

(LOAD ‘EXPORTS.ALL).

Then, to begin working on items in the TEDIT source, I would do (LOADFROM ’TEDIT) so that that the TEDIT source file and its contents are “noticed” by the running system.

A safer way to work on low-level system files is to first LOAD(SYSEDIT).  This gets the EXPORTS.ALL, but also sets some flags so that functions that you edit don’t automatically get installed as running code.  That protects you, for example, from accidentally redefining CONS.

In short, it could be that your (LOAD ’SKETCH) might have worked if you had first done (LOAD ‘EXPORTS.ALL) or LOAD(SYSTEDIT). But again, that would only be because you wanted to work on SKETCH, not just use it.

Hope this helps.

**Comment by pamoroso** [comment]
Thanks for the explanation, I just wanted to run Sketch as a user and not work on it.

**Comment by pamoroso** [comment]
By the way, I assumed that, if both the source and the compiled version of a system were available, Interlisp would load the latter by default.

**Comment by nbriggs** [comment]
If you were to `(FILESLOAD SKETCH)` it would choose the compiled file if available, otherwise the source.  That way you also don't have to think about whether it's an `LCOM` or a `DFASL` compiled file.  See the manual for a description of the options for `FILESLOAD`.

**Comment by pamoroso** [comment]
Very interesting, thanks all. If there are no further comment's I'll close the issue.

**Comment by masinter** [comment]
Anything that requires as much explanation is an opportunity for documentation

**Comment by pamoroso** [comment]
Right, it's a great opportunity and I'll put together a PR based on this discussion. A change to the project site that may go under Documentation > Medley > [Orientation to Medley](https://interlisp.org/docs/medley/orientation/), perhaps inside or near section "Saving & Loading Source Code To/From Disk Files".

**Comment by pamoroso** [comment]
Doesn't `LOAD` also choose the compiled file if available? From the IRM entry for `LOAD`:

> Reads successive expressions from `FILE` (with `FILERDTBL` as read table) and evaluates each as it is read, until it reads either `NIL`, or the single atom `STOP`. Note that `LOAD` can be used to load both symbolic and compiled files.

I'm not sure I get the advantages of `FILESLOAD` over `LOAD` when loading a file for execution rather than for editing the source.

**Comment by nbriggs** [comment]
No, `LOAD` loads exactly what you tell it to load, which may be *either* a symbolic or a compiled file, depending on what you tell it.

If you're putting things in your personal (or site) init file because you want the package always loaded, then using a FILESLOAD expression is what you'd do.  You might also use other options, like `(FILESLOAD (SYSLOAD FROM LISPUSERSDIRECTORIES) FOO)`.   With FILESLOAD you don't have to know whether it's going to be an LCOM or a DFASL compiled file, etc., see the FILESLOAD documentation.

**Comment by pamoroso** [comment]
Got it, it's either source or compiled.

**Comment by masinter** [comment]
I marked this as "documentation"; perhaps it belongs with the other usage notes from @sethm and the notes originally from @blakemcbride. We're also doing some introductory videos/youtube, but my feeling is that everything in the videos should (eventually) also be covered by the documentation.

@skaisler1 's books could also play a part... 

It's also possible to change LOAD (and CL:LOAD) to do the "right thing", especially if we gave Interlisp source files an extension (a change we've discussed in other issues #265 #497 #15).
Then if you said (LOAD "sketch") it could do the right thing because it would look for sketch.lcom sketch.dfasl sketch.ilisp because the distinction between "load the source" and "load compiled" and "load the best" would be clearer.

**Comment by rmkaplan** [comment]
One issue is what the target of the load operation should be, source or compiled.  An orthogonal issue is “how” it should be brought it, which is determined by other options to FILESLOAD or by the LDFLG argument of the more primitive LOAD function.

SYSLOAD says that the file (presumaby the compiled file if it exists) is the target, everything from it should be brought into memory… and then the file itself should be mostly forgotten in the sense that it is not registered  with (“noticed by”) the file package.  Just like the fileCOMS for all the system files are not preserved in the makesys, filepkg-level information about that file is also not preserved. You get the stuff, but you don’t remember where the stuff came from (except for what the separate where-is database might have recorded).

The file can also be loaded PROP, which means that everything is brought in and the file is also registered/noticed.  But in this mode the source-definitions of functions are saved on properly lists and do not override existing active definitions (unless/until other operations that you do cause those definitions to be “unsaved”).

And LOADFROM is a way of registering the file for future incremental file-package operations but without yet really loading anything from it. There are a few other options.

This is all about managing the relationships between the information stored externally in files and how/whether that information takes up residence in the running memory image.

**Comment by nbriggs** [comment]
I strongly disagree with making LOAD do the search for compiled or source. If I wanted that I’d call FILESLOAD, and FILESLOAD (and various others, possibly) rely on LOAD as the low level implementation, and there would be many as yet unrecognized implications to making that change (as well as invalidating all the existing documentation)

**Comment by masinter** [comment]
The idea would be if you say (load "foo") but there's only "foo.ilisp" and "foo.dfasl".
More likely would be to survey current CL implementations and change CL:LOAD to match.
The CL spec 
```
man cl:load
```

only says "implementation dependent".

**Comment by nbriggs** [comment]
I expect `(IL:LOAD "foo")` to produce a FILE-NOT-FOUND error if only "foo.ilisp" and "foo.dfasl" exist (in the connected directory).  If you want the search behavior then `(IL:FILESLOAD FOO)` is the way to express it.

cl:load is a different matter.  I don't know what the system currently documents, but again, if we're going to change it, we need to be able to fix and regenerate the documentation and check any uses to make sure we're not breaking things.

Stuff getting inadvertently broken and not discovered until much later is already a problem we're experiencing, and changing things like this without tests and sufficient people who are willing and competent to vet the changes is just going to lead to chaos.

**Comment by Michele31415** [comment]
The way I always used it back in the day was that if I wanted the source 
I'd simply say LOAD (FOO) and if I wanted the compiled version I'd say 
LOAD (FOO.LCOM).  It seems like it's getting awfully complicated.

             - Michele
